### PR TITLE
[Slider] Modernize implementation

### PR DESCRIPTION
**Pull request by mj12albert**
### Summary

1. Replaced `useCompound` with `Composite`

Previously the `id` attribute of all the `input` elements were passed to `useCompound` as metadata to generate the `for` attribute of the `output` element.

However the refs provided by Composite can't be used during the render of `Output` for this, so the `id`s have to be held in a new state

2. Converted types to namespaces
3. Fix https://github.com/mui/base-ui/issues/635
4. Hooks/context are no longer exported for now - it's no longer possible to (easily) implement a Material Design slider as it would require custom subcomponents
5. Matches latest export convention in https://github.com/mui/base-ui/pull/700

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/base-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

File: docs/app/experiments/slider-marks.tsx [hunk 1]
```diff
@@ -1,7 +1,7 @@
 'use client';
 import * as React from 'react';
 import * as Slider from '@base_ui/react/Slider';
-import { useSliderContext } from '@base_ui/react/Slider';
+import { useSliderContext } from '../../../packages/mui-base/src/Slider/Root/SliderContext';
```
Comment (michaldudak) [inline docs/app/experiments/slider-marks.tsx:4]: If there is a legitimate use case for the context, we should export it from the package.

File: docs/data/components/slider/UnstyledSliderIntroduction/css-modules/index.tsx [hunk 2]
```diff
@@ -12,159 +13,36 @@ export default function UnstyledSliderIntroduction() {
...
-        <Slider.Output className="Slider-output" />
-        <Slider.Control className="Slider-control">
-          <Slider.Track className="Slider-track">
-            <Slider.Indicator className="Slider-indicator" />
-            <Slider.Thumb className="Slider-thumb" />
+        <Slider.Output className={classes.output} />
+        <Slider.Control className={classes.control}>
+          <Slider.Track className={classes.track}>
+            <Slider.Indicator className={classes.indicator} />
+            <Slider.Thumb className={classes.thumb} inputId=":slider-thumb-input:" />
```
Comment (michaldudak) [inline docs/data/components/slider/UnstyledSliderIntroduction/css-modules/index.tsx:31]: The `inputId` prop is not documented

File: packages/mui-base/src/Slider/Control/SliderControl.tsx [hunk 3]
```diff
@@ -71,6 +71,12 @@ const SliderControl = React.forwardRef(function SliderControl(
   return renderElement();
 });
 
+export namespace SliderControl {
+  export interface Props extends BaseUIComponentProps<'span', SliderRoot.OwnerState> {}
```
Comment (michaldudak) [inline packages/mui-base/src/Slider/Control/SliderControl.tsx:75]:
```suggestion
  export interface OwnerState extends SliderRoot.OwnerState {}
  
  export interface Props extends BaseUIComponentProps<'span', OwnerState> {}
```

+ other cases where the OwnerState isn't defined explicitly

File: packages/mui-base/src/Slider/Indicator/useSliderIndicator.ts [hunk 4]
```diff
@@ -21,9 +19,19 @@ const axisProps = {
   },
 };
 
-function useSliderIndicator(
-  parameters: UseSliderIndicatorParameters,
-): UseSliderIndicatorReturnValue {
+/**
```
Comment (michaldudak) [inline packages/mui-base/src/Slider/Indicator/useSliderIndicator.ts:22]: Hooks are not documented, so this header is invalid. The current docs:api script doesn't create it anymore.

File: packages/mui-base/src/Slider/Root/SliderContext.tsx [hunk 5]
```diff
@@ -0,0 +1,20 @@
+'use client';
+import * as React from 'react';
+import type { SliderRoot } from './SliderRoot';
+
+/**
+ * @ignore - internal component.
+ */
+export const SliderContext = React.createContext<SliderRoot.Context | undefined>(undefined);
```
Comment (michaldudak) [inline packages/mui-base/src/Slider/Root/SliderContext.tsx:8]: Context type should be defined in this file (and be called `SliderContext`)

Comment (mj12albert) [reply packages/mui-base/src/Slider/Root/SliderContext.tsx]: Updated ~ I've also renamed it to `SliderRootContext`/`useSliderRootContext` to match other components

File: docs/data/components/slider/UnstyledSliderIntroduction/css-modules/index.tsx (hunk 2 above)
Comment (mj12albert) [reply docs/data/components/slider/UnstyledSliderIntroduction/css-modules/index.tsx:31]: Updated!

File: docs/app/experiments/slider-marks.tsx (hunk 1 above)
Comment (mj12albert) [reply docs/app/experiments/slider-marks.tsx:4]: We discussed this and decided to revisit exporting Context(s) after the alpha phase

File: packages/mui-base/src/Slider/Root/useSliderRoot.ts [hunk 6]
```diff
@@ -190,10 +188,31 @@ function useSliderRoot(parameters: UseSliderParameters): UseSliderReturnValue {
...
+      if (index > -1 && inputId !== undefined) {
+        setInputMap((prevMap) => new Map(prevMap).set(index, inputId));
+      }
+
+      return {
+        deregister: deregisterInputId,
+      };
+    },
+    [deregisterInputId],
+  );
```
Comment (atomiks) [inline packages/mui-base/src/Slider/Root/useSliderRoot.ts:215]: Is it possible to modify the `Composite` source to allow for metadata instead?

Comment (mj12albert) [reply packages/mui-base/src/Slider/Root/useSliderRoot.ts:215]: Should be! How about doing this when refactoring Tabs to use Composite? Tabs uses metadata more extensively, and I wanted to get this PR merged ASAP to unblock the ESM work @michaldudak is doing 🙏

Comment (atomiks) [reply packages/mui-base/src/Slider/Root/useSliderRoot.ts:215]: Seems ok to me as of now

File: packages/mui-base/src/Slider/Root/SliderContext.tsx (hunk 5 above)
Comment (michaldudak) [reply packages/mui-base/src/Slider/Root/SliderContext.tsx:8]: Yeah, that's actually what I meant 😅

**Review by michaldudak — APPROVED** [review]

**Pull request merged**